core/Updates/3.0.0-b2.php

```diff
@@ -21,4 +23,84 @@ public static function isMajorUpdate()
     {
         return true;
     }
+
+    /**
+     * @var MigrationFactory
+     */
+    private $migration;
+
+    /**
+     * @var string
+     */
+    private $userTable = 'user';
+
+    public function __construct(MigrationFactory $factory)
+    {
+        $this->migration = $factory;
+    }
+
+    /**
+     * Here you can define one or multiple SQL statements that should be executed during the update.
+     * @param Updater $updater
+     * @return Migration[]
+     */
+    public function getMigrations(Updater $updater)
+    {
+        $migrations = [];
+        $migrations = $this->getUserDatabaseMigrations($migrations);
+
+        return $migrations;
+    }
+
+
+    public function doUpdate(Updater $updater)
+    {
+        $updater->executeMigrations(__FILE__, $this->getMigrations($updater));
+
+        $this->hashUserPasswords();
+    }
+
+      /**
+     * @param Migration[] $queries
+     * @return Migration[]
+     */
+    private function getUserDatabaseMigrations($queries)
+    {
+        $queries[] = $this->migration->db->changeColumn($this->userTable, 'password', 'password', 'VARCHAR(255) NOT NULL');
+        $queries[] = $this->migration->db->addColumn($this->userTable, 'password_hashed', 'VARCHAR(255) NOT NULL', 'password');
```

Comment:
+1 to no longer regenerate token when password changes but then we need to have a feature in the UI to generate a new token. I'd say we could just create a new issue for this to not generate a new token when password changes to keep things easy and doable for now. I'd still include the username in the generation but password not needed anymore indeed.

> Regarding the passwords I also thought about just applying bcrypt around already hashed MD5 passwords. 
I really like that idea. In Piwik 4 we could then still think about removing the md5($password) and breaking the API.